internal/configv2/configv2.go

```diff
@@ -0,0 +1,511 @@
+	"github.com/census-instrumentation/opencensus-service/internal/factories"
+)
+
+// These are errors that can be returned by Load(). Note that error codes are not part
+// of Load()'s public API, they are for internal unit testing only.
+type configErrorCode int
+
+const (
+	_ configErrorCode = iota // skip 0, start errors codes from 1.
+	errInvalidTypeAndNameKey
+	errUnknownReceiverType
+	errUnknownExporterType
+	errUnknownProcessorType
+	errInvalidPipelineType
+	errDuplicateReceiverName
+	errDuplicateExporterName
+	errDuplicateProcessorName
+	errDuplicatePipelineName
+	errMissingPipelines
+	errPipelineMustHaveReceiver
+	errPipelineMustHaveExporter
+	errPipelineMustHaveProcessors
+	errPipelineReceiverNotExists
+	errPipelineProcessorNotExists
+	errPipelineExporterNotExists
+	errMetricPipelineCannotHaveProcessors
+	errUnmarshalError
+)
+
+type configError struct {
+	msg  string          // human readable error message.
+	code configErrorCode // internal error code.
+}
+
+func (e *configError) Error() string {
+	return e.msg
+}
+
+// receiversKeyName is the configuration key name for receivers section.
+const receiversKeyName = "receivers"
+
+// exportersKeyName is the configuration key name for exporters section.
+const exportersKeyName = "exporters"
+
+// processorsKeyName is the configuration key name for processors section.
+const processorsKeyName = "processors"
+
+// pipelinesKeyName is the configuration key name for pipelines section.
+const pipelinesKeyName = "pipelines"
+
+// typeAndNameSeparator is the separator that is used between type and name in type/name composite keys.
+const typeAndNameSeparator = "/"
+
+const (
+	tracesDataTypeStr  = "traces"
+	metricsDataTypeStr = "metrics"
+)
+
+// Load loads a ConfigV2 from Viper.
+func Load(v *viper.Viper) (*configmodels.ConfigV2, error) {
+
+	var config configmodels.ConfigV2
+
+	// Load the config.
+
+	receivers, err := loadReceivers(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Receivers = receivers
+
+	exporters, err := loadExporters(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Exporters = exporters
+
+	processors, err := loadProcessors(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Processors = processors
+
+	pipelines, err := loadPipelines(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Pipelines = pipelines
+
+	// Config is loaded. Now validate it.
+
+	if err := validateConfig(&config); err != nil {
+		return nil, err
+	}
+
+	return &config, nil
+}
+
+// decodeTypeAndName decodes a key in type[/name] format into type and fullName.
+// fullName is the key normalized such that type and name components have spaces trimmed.
+// The "type" part must be present, the forward slash and "name" are optional.
+func decodeTypeAndName(key string) (typeStr, fullName string, err error) {
+	items := strings.SplitN(key, typeAndNameSeparator, 2)
+	if len(items) < 1 {
+		// We need at least one part (the type).
+		err = errors.New("Invalid key")
```

Comment:
nit: the comment has better information than the error message.

Reply:
Fixed.